pkg/services/pluginsintegration/clientmiddleware/team_headers_middleware.go

```diff
@@ -0,0 +1,181 @@
+	return plugins.ClientMiddlewareFunc(func(next plugins.Client) plugins.Client {
+		return &TeamHeadersMiddleware{
+			next: next,
+		}
+	})
+}
+
+type TeamHeadersMiddleware struct {
+	next plugins.Client
+}
+
+type TeamHeadersJSONData struct {
+	TeamHeaders TeamHttpHeaders `json:"teamHeaders"`
+}
+
+type TeamHttpHeaders map[string][]TeamHttpHeader
+
+type TeamHttpHeader struct {
+	Header string `json:"header"`
+	Value  string `json:"value"`
+}
+
+func (m *TeamHeadersMiddleware) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
+	if req == nil {
+		return m.next.QueryData(ctx, req)
+	}
+
+	err := m.setHeaders(ctx, req.PluginContext, req)
+	if err != nil {
+		return nil, err
+	}
+
+	return m.next.QueryData(ctx, req)
+}
+
+func (m *TeamHeadersMiddleware) CallResource(ctx context.Context, req *backend.CallResourceRequest, sender backend.CallResourceResponseSender) error {
+	if req == nil {
+		return m.next.CallResource(ctx, req, sender)
+	}
+
+	err := m.setHeaders(ctx, req.PluginContext, req)
+	if err != nil {
+		return err
+	}
+
+	return m.next.CallResource(ctx, req, sender)
+}
+
+func (m *TeamHeadersMiddleware) CheckHealth(ctx context.Context, req *backend.CheckHealthRequest) (*backend.CheckHealthResult, error) {
+	if req == nil {
+		return m.next.CheckHealth(ctx, req)
+	}
+
+	err := m.setHeaders(ctx, req.PluginContext, req)
+	if err != nil {
+		return nil, err
+	}
+
+	return m.next.CheckHealth(ctx, req)
+}
+
+func (m *TeamHeadersMiddleware) CollectMetrics(ctx context.Context, req *backend.CollectMetricsRequest) (*backend.CollectMetricsResult, error) {
+	return m.next.CollectMetrics(ctx, req)
+}
+
+func (m *TeamHeadersMiddleware) SubscribeStream(ctx context.Context, req *backend.SubscribeStreamRequest) (*backend.SubscribeStreamResponse, error) {
+	return m.next.SubscribeStream(ctx, req)
+}
+
+func (m *TeamHeadersMiddleware) PublishStream(ctx context.Context, req *backend.PublishStreamRequest) (*backend.PublishStreamResponse, error) {
+	return m.next.PublishStream(ctx, req)
+}
+
+func (m *TeamHeadersMiddleware) RunStream(ctx context.Context, req *backend.RunStreamRequest, sender *backend.StreamSender) error {
+	return m.next.RunStream(ctx, req, sender)
+}
+
+func (m *TeamHeadersMiddleware) setHeaders(ctx context.Context, pCtx backend.PluginContext, req interface{}) error {
+	reqCtx := contexthandler.FromContext(ctx)
+	// if request not for a datasource or no HTTP request context skip middleware
+	if req == nil || pCtx.DataSourceInstanceSettings == nil || reqCtx == nil || reqCtx.Req == nil {
+		return nil
+	}
+
+	settings := pCtx.DataSourceInstanceSettings
+	jsonDataBytes, err := simplejson.NewJson(settings.JSONData)
+	if err != nil {
+		return err
+	}
+
+	ds := &datasources.DataSource{
+		ID:       settings.ID,
+		OrgID:    pCtx.OrgID,
+		JsonData: jsonDataBytes,
+		Updated:  settings.Updated,
+	}
+
+	// TODO: add teams to User struct in grafana-plugin-sdk-go@v0.179.0/backend/common.go
```

Comment:
`pCtx.User` does not contain teams list, so we should extend it here
https://github.com/grafana/grafana/blob/bdeb829cf6fdafbe2b6ba6dcaa1ce6c5bc637cdb/pkg/services/pluginsintegration/adapters/adapters.go#L48

Reply:
Please see code snippet I sent offline - shows how to get a hold of the user/teams from context

Reply:
Yeah, sorry, I didn't notice it, we definitely do not need to extend User struct in sdk since we can get user identity from request context.